src/MSALWrapper/TokenFetcher.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.MSALWrapper
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.Authentication.MSALWrapper.AuthFlow;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// A functional orchestrator of doing auth using the building blocks
+    /// of <see cref="AuthFlowFactory"/> and <see cref="AuthFlowExecutor"/>.
+    /// </summary>
+    public static class TokenFetcher
+    {
+        /// <summary>
+        /// The result of running <see cref="TokenFetcher"/>.
+        /// </summary>
+        public record Result
+        {
+            /// <summary>
+            /// Gets the success <see cref="AuthFlowResult"/> from <see cref="Attempts"/> if one exists, null otherwise.
+            /// </summary>
+            public AuthFlowResult Success => this.Attempts.FirstOrDefault(result => result.Success);
+
+            /// <summary>
+            /// Gets all the attempts made to authenticate.
+            /// </summary>
+            public List<AuthFlowResult> Attempts { get; init; }
+        }
+
+        /// <summary>
+        /// Run the authenticatuion process.
+        /// </summary>
+        /// <param name="logger">A logger.</param>
+        /// <param name="client">Client.</param>
+        /// <param name="tenant">Tenant.</param>
+        /// <param name="scopes">Scopes.</param>
+        /// <param name="mode">Modes.</param>
+        /// <param name="domain">Domain.</param>
+        /// <param name="prompt">Prompt Hint.</param>
+        /// <param name="timeout">Timeout.</param>
+        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
+        public static async Task<Result> AccessTokenAsync(ILogger logger, Guid client, Guid tenant, IEnumerable<string> scopes, AuthMode mode, string domain, string prompt, TimeSpan timeout)
+        {
+            var authFlows = AuthFlowFactory.Create(

```

Comment:
This method is almost copied directly from `CommandAad.GetToken` with 1 important change - there's not yet a way to inject an IAuthFlow. We are tightly coupled with the AuthFlowFactory.

There is only 1 test requireing that and it's sort of an integration level test as all the components used here, aside from the lock, are unit tested already. I'm still pondering if we will want to add this ability. I'm leaning yes - and will do that when required to do by updating the CommandAad usage.